safehttp/form_test.go

```diff
@@ -0,0 +1,1051 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package safehttp
+
+import (
+	"errors"
+	"fmt"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/safehtml"
+	"github.com/google/safehtml/template"
+	"math"
+	"net/http"
+	"net/http/httptest"
+	"reflect"
+	"strconv"
+	"strings"
+	"testing"
+)
+
+const (
+	status200OK     = "200 OK"
+	status400BadReq = "400 Bad Request"
+)
+
+type dispatcher struct{}
+
+func (d *dispatcher) Write(rw http.ResponseWriter, resp Response) error {
+	switch x := resp.(type) {
+	case safehtml.HTML:
+		_, err := rw.Write([]byte(x.String()))
+		return err
+	default:
+		panic("not a safe response type")
+	}
+}
+
+func (d *dispatcher) ExecuteTemplate(rw http.ResponseWriter, t Template, data interface{}) error {
+	switch x := t.(type) {
+	case *template.Template:
+		return x.Execute(rw, data)
+	default:
+		panic("not a safe response type")
+	}
+}
+
+// A helper function that returns a parsed Form or an empty Form and any errors
+// that occured during parsing
+func getParsedForm(r *IncomingRequest) (*Form, error) {
+	if r.req.Method == "GET" {
+		f, err := r.QueryForm()
+		return f, err
+	}
+
+	if !strings.HasPrefix(r.req.Header.Get("Content-Type"), "multipart/form-data") {
+		f, err := r.PostForm()
+		return f, err
+	}
+	mf, err := r.MultipartForm(32 << 20)
+	return &mf.Form, err
+}
+
+func TestFormValidInt(t *testing.T) {
+	maxIntToString := strconv.FormatInt(math.MaxInt64, 10)
```

Comment:
It's not a function or anything like that, so the name `xToY` doesn't really fit.